src/Common/src/TypeSystem/IL/Stubs/ComparerIntrinsics.cs

```diff
@@ -0,0 +1,216 @@
+        /// </summary>
+        public static MethodIL EmitComparerCreate(MethodDesc target)
+        {
+            return EmitComparerAndEqualityComparerCreateCommon(target, "Comparer", "IComparable`1");
+        }
+
+        /// <summary>
+        /// Generates a specialized method body for EqualityComparer`1.Create or returns null if no specialized body can be generated.
+        /// </summary>
+        public static MethodIL EmitEqualityComparerCreate(MethodDesc target)
+        {
+            return EmitComparerAndEqualityComparerCreateCommon(target, "EqualityComparer", "IEquatable`1");
+        }
+
+        /// <summary>
+        /// Gets the concrete type EqualityComparer`1.Create returns or null if it's not known at compile time.
+        /// </summary>
+        public static TypeDesc GetEqualityComparerForType(TypeDesc comparand)
+        {
+            return GetComparerForType(comparand, "EqualityComparer", "IEquatable`1");
+        }
+
+        private static MethodIL EmitComparerAndEqualityComparerCreateCommon(MethodDesc methodBeingGenerated, string flavor, string interfaceName)
+        {
+            // We expect the method to be fully instantiated
+            Debug.Assert(!methodBeingGenerated.IsTypicalMethodDefinition);
+
+            TypeDesc owningType = methodBeingGenerated.OwningType;
+            TypeDesc comparedType = owningType.Instantiation[0];
+
+            // If the type is canonical, we use the default implementation provided by the class library.
+            // This will rely on the type loader to load the proper type at runtime.
+            if (comparedType.IsCanonicalSubtype(CanonicalFormKind.Any))
+                return null;
+
+            TypeDesc comparerType = GetComparerForType(comparedType, flavor, interfaceName);
+            Debug.Assert(comparerType != null);
+
+            ILEmitter emitter = new ILEmitter();
+            var codeStream = emitter.NewCodeStream();
+
+            codeStream.Emit(ILOpcode.newobj, emitter.NewToken(comparerType.GetParameterlessConstructor()));
+            codeStream.Emit(ILOpcode.dup);
+            codeStream.Emit(ILOpcode.stsfld, emitter.NewToken(owningType.GetKnownField("_default")));
+            codeStream.Emit(ILOpcode.ret);
+
+            return emitter.Link(methodBeingGenerated);
+        }
+
+        /// <summary>
+        /// Gets the comparer type that is suitable to compare instances of <paramref name="type"/>
+        /// or null if such comparer cannot be determined at compile time.
+        /// </summary>
+        private static TypeDesc GetComparerForType(TypeDesc type, string flavor, string interfaceName)
+        {
+            TypeSystemContext context = type.Context;
+
+            if (context.IsCanonicalDefinitionType(type, CanonicalFormKind.Any) ||
+                (type.IsRuntimeDeterminedSubtype && !type.HasInstantiation))
```

Comment:
So this will optimize the `struct X<T> : IEquatable<T>` case you have been worried about properly, correct?

Reply:
Yes. This method handles it, because it's shared with the logic that ensures templates are presents (and the devirtualization logic). But if T ends up being Canon, we reject generating a specialized method body for it in `EmitComparerAndEqualityComparerCreateCommon` and fall back to the type loader.